README.md
# rstgroup/statsd-module [![Build Status](https://travis-ci.org/rstgroup/statsd-module.svg?branch=master)](https://travis-ci.org/rstgroup/statsd-module)

Module is based on [domnikl/statsd-php](https://github.com/domnikl/statsd-php).


## Installation

Update your composer.json with following lines:

```javascript
"require": {
    "rstgroup/statsd-module": "^1.0"
}
```


## Configuration

In your ZF2 application config add to module list

```php
return [
    'modules' => [
        'RstGroup\\StatsdModule',
    ],
];
```

also in your autoload config based on enviroment add statsd client configuration

```php
return [
    'statsd' => [
        'tcp' => [
            'host' => 'example_host',
            'port' => example_port,
            'timeout' => null,
            'persistent' => false,
            'mtu' => 1500,
        ],
        'udp' => [
            'host' => 'example_host',
            'port' => example_port,
            'timeout' => null,
            'persistent' => false,
            'mtu' => 1500,
        ],
        'namespace' => 'services.default',
        'connectionType' => 'RstGroup\Statsd\Connection\Blackhole',
    ],
];
```

Connection types:

* `RstGroup\Statsd\Connection\Udp`
* `RstGroup\Statsd\Connection\Tcp` errors will not be suppressed in this mode
* `RstGroup\Statsd\Connection\Memory` in this mode sending is disabled, but messages are collected
* `RstGroup\Statsd\Connection\Blackhole` in this mode sending is disabled




## Usage

Typical usage from ServiceManager

```php
$client = $serviceManager->get("RstGroup\Statsd\Client");
```


CHANGELOG.md
# 1.2.0

* ZF3 support added

# 1.1.0

* Require only required zend packages

# 1.0.0

* First release


tests/unit/phpunit/bootstrap.php
<?php

$autoload = __DIR__ . '/../../../vendor/autoload.php';

if (file_exists($autoload)) {
    require_once $autoload;
} else {
    echo "Missing autoload file. Did you run `composer`?";
    exit(0);
}


tests/unit/phpunit/ClientFactoryTest.php
<?php

namespace RstGroup\StatsdModule\Tests\Unit\PHPUnit\ClientFactory;

use Domnikl\Statsd\Client;
use Interop\Container\ContainerInterface;
use RstGroup\StatsdModule\ClientFactory;

class ClientFactoryTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Testing proper building statsd client
     *
     * @dataProvider connectionProvider
     * @param array $config
     */
    public function testWillReturnProperStatsdClientInstance(array $config)
    {
        $factory = new ClientFactory();

        $containerMock = $this->getMock(ContainerInterface::class);

        $connectionMock = $this->getMockBuilder($config['realConnection'])
            ->disableOriginalConstructor()
            ->getMock();

        $serviceLocatorServices = [
            'Config' => $config,
            $config['statsd']['connectionType'] => $connectionMock,
        ];

        $containerMock->expects($this->any())
            ->method('get')
            ->will($this->returnCallback(function($name) use ($serviceLocatorServices) {
                return $serviceLocatorServices[$name];
            }));

        $this->assertInstanceOf(Client::class, $factory($containerMock));
    }

    public function connectionProvider()
    {
        return [
            [
                [
                    'statsd' => [
                        'namespace' => 'services.default',
                        'connectionType' => 'RstGroup\Statsd\Connection\Blackhole',
                    ],
                    'realConnection' => 'Domnikl\Statsd\Connection\Blackhole',
                ],
            ],
            [
                [
                    'statsd' => [
                        'namespace' => 'services.default',
                        'connectionType' => 'RstGroup\Statsd\Connection\Tcp',
                    ],
                    'realConnection' => 'Domnikl\Statsd\Connection\TcpSocket',
                ],
            ],
            [
                [
                    'statsd' => [
                        'namespace' => 'services.default',
                        'connectionType' => 'RstGroup\Statsd\Connection\Udp',
                    ],
                    'realConnection' => 'Domnikl\Statsd\Connection\UdpSocket',
                ],
            ],
            [
                [
                    'statsd' => [
                        'namespace' => 'services.default',
                        'connectionType' => 'RstGroup\Statsd\Connection\Memory',
                    ],
                    'realConnection' => 'Domnikl\Statsd\Connection\InMemory',
                ],
            ],
        ];
    }
}


tests/unit/phpunit/Connection/TcpFactoryTest.php
<?php

namespace RstGroup\StatsdModule\Tests\Unit\PHPUnit\TcpFactory;

use Domnikl\Statsd\Connection\TcpSocket;
use Interop\Container\ContainerInterface;
use RstGroup\StatsdModule\Connection\TcpFactory;

class TcpFactoryTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Testing proper building tcp connection
     */
    public function testWillReturnProperlyConfiguredTcpConnectorForStatsdClient()
    {
        $config = [
            'statsd' => [
                'tcp'            => [
                    'host'       => 'localhost',
                    'port'       => 8125,
                    'timeout'    => null,
                    'persistent' => false,
                    'mtu'        => 1500,
                ],
            ],
        ];

        $factory = new TcpFactory();

        $containerMock = $this->getMock(ContainerInterface::class);
        $containerMock->expects($this->once())
                           ->method('get')
                           ->with($this->equalTo('Config'))
                           ->will($this->returnValue($config));

        $connection = $factory($containerMock);
        $this->assertInstanceOf(TcpSocket::class, $connection);
        $this->assertEquals($connection->getHost(), $config['statsd']['tcp']['host']);
        $this->assertEquals($connection->getPort(), $config['statsd']['tcp']['port']);
        $this->assertEquals($connection->getTimeout(), $config['statsd']['tcp']['timeout']);
    }
}


config/module.config.php
<?php

return [
    'service_manager' => [
        'factories' => [
            'RstGroup\Statsd\Client' => 'RstGroup\StatsdModule\ClientFactory',
            'RstGroup\Statsd\Connection\Udp' => 'RstGroup\StatsdModule\Connection\UdpFactory',
            'RstGroup\Statsd\Connection\Tcp' => 'RstGroup\StatsdModule\Connection\TcpFactory',
        ],
        'invokables' => [
            'RstGroup\Statsd\Connection\Blackhole' => 'Domnikl\Statsd\Connection\Blackhole',
            'RstGroup\Statsd\Connection\Memory' => 'Domnikl\Statsd\Connection\InMemory',
        ],
    ],
    // Statsd client module start configuration
    'statsd' => [
        'tcp' => [
            'host' => 'localhost',
            'port' => 8125,
            'timeout' => null,
            'persistent' => false,
            'mtu' => 1500,
        ],
        'udp' => [
            'host' => 'localhost',
            'port' => 8125,
            'timeout' => null,
            'persistent' => false,
            'mtu' => 1500,
        ],
        'namespace' => 'services.default',
        'connectionType' => 'RstGroup\Statsd\Connection\Blackhole', // service name from service manager
    ],
];
